src/pymor/parameters/functionals.py

```diff
@@ -105,16 +125,47 @@ class ExpressionParameterFunctional(GenericParameterFunctional):
     functions['polar'] = lambda x: (np.linalg.norm(x, axis=-1), np.arctan2(x[..., 1], x[..., 0]) % (2*np.pi))
     functions['np'] = np
 
-    def __init__(self, expression, parameter_type, name=None):
-        self.expression = expression
-        code = compile(expression, '<expression>', 'eval')
-        functions = self.functions
-        mapping = lambda mu: eval(code, functions, mu)
-        super().__init__(mapping, parameter_type, name)
+    def __init__(self, expression, parameter_type, name=None, derivative_expressions=None):
+        if derivative_expressions is None:
+            self.expression = expression
+            code = compile(expression, '<expression>', 'eval')
+            functions = self.functions
+            mapping = lambda mu: eval(code, functions, mu)
+            super().__init__(mapping, parameter_type, name)
+            self.__auto_init(locals())
+        else:
+            self.expression = expression
+            self.derivative_expressions = derivative_expressions
+            functions = self.functions
+            code_exp = compile(expression, '<expression>', 'eval')
+            exp_mapping = lambda mu: eval(code_exp, functions, mu)
+            derivative_mappings = {}
+            def get_lambda(exp_code):
+                return lambda mu: eval(exp_code, functions, mu)
+            for (key,exp) in derivative_expressions.items():
+                exp_array = np.array(exp, dtype=object)
+                mappings = np.empty(exp_array.shape, dtype=object)
+                if exp_array.shape == ():
+                    exp_code = compile(exp, '<expression>', 'eval')
+                    mapping = get_lambda(exp_code)
+                    mappings = np.array(mapping)
+                elif np.shape(exp_array.shape) == (1,):
```

Comment:
Parameter components in pyMOR can be arbitrary numpy arrays, so it is not ideal to restrict to 0D,1D and 2D arrays. There are numpy methods which allow you to iterate over all elements of an arbitrary numpy array.